phoenix-core/src/main/java/org/apache/phoenix/index/GlobalIndexRowVerifier.java

```diff
@@ -0,0 +1,208 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.phoenix.index;
+
+import static org.apache.phoenix.coprocessor.BaseScannerRegionObserver.CHECK_VERIFY_COLUMN;
+import static org.apache.phoenix.coprocessor.BaseScannerRegionObserver.EMPTY_COLUMN_FAMILY_NAME;
+import static org.apache.phoenix.coprocessor.BaseScannerRegionObserver.PHYSICAL_DATA_TABLE_NAME;
+import static org.apache.phoenix.query.QueryConstants.GLOBAL_INDEX_VERIFIED_COLUMN_NAME_BYTES;
+import static org.apache.phoenix.schema.types.PDataType.FALSE_BYTES;
+import static org.apache.phoenix.schema.types.PDataType.TRUE_BYTES;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import org.apache.hadoop.hbase.Cell;
+import org.apache.hadoop.hbase.DoNotRetryIOException;
+import org.apache.hadoop.hbase.client.Delete;
+import org.apache.hadoop.hbase.client.Mutation;
+import org.apache.hadoop.hbase.client.ResultScanner;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.hadoop.hbase.client.Table;
+import org.apache.hadoop.hbase.coprocessor.RegionCoprocessorEnvironment;
+import org.apache.hadoop.hbase.io.ImmutableBytesWritable;
+import org.apache.hadoop.hbase.regionserver.Region;
+import org.apache.hadoop.hbase.regionserver.RegionScanner;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.apache.phoenix.coprocessor.BaseScannerRegionObserver;
+import org.apache.phoenix.util.EnvironmentEdgeManager;
+import org.apache.phoenix.util.IndexUtil;
+import org.apache.phoenix.util.SchemaUtil;
+import org.apache.phoenix.util.ServerUtil;
+
+/**
+ * 
+ * Class that verifies a given row of a non-transactional global index.
+ * An instance of this class is created by a scan region observer
+ * and called to verify individual rows of a non-transactional global index
+ *
+ */
+public class GlobalIndexRowVerifier {
+    private final long AGE_THRESHOLD_TO_DELETE_MS = 3600*1000; /* One hour */
+    private boolean verify;
+    private Scan scan;
+    private Scan indexScan;
+    private Scan buildIndexScan = null;
+    private Table dataHTable;
+    private byte[] emptyCF;
+    private IndexMaintainer indexMaintainer = null;
+    private byte[][] viewConstants = null;
+    private RegionCoprocessorEnvironment env;
+    private Region region = null;
+
+    public GlobalIndexRowVerifier(RegionCoprocessorEnvironment env, Scan scan) {
+        verify = scan.getAttribute(CHECK_VERIFY_COLUMN) != null;
+        if (verify) {
+            this.env = env;
+            this.scan = scan;
+            emptyCF = scan.getAttribute(EMPTY_COLUMN_FAMILY_NAME);
+        }
+    }
+
+    public GlobalIndexRowVerifier(RegionCoprocessorEnvironment env, Scan scan,
```

Comment:
Would be good to have a test class for this class with tests that verify:

1. Calls to verifyRow with the verified bytes set to true returns the original row
2. Calls to verifyRow with the verified bytes set to false returns the data row (can probably use mocks for this)
3. Calls to verifyRow with the verified bytes missing returns the data row (likewise can probably use mocks)
4. Verify that Cases 2 and 3 call rebuildSingleIndexRow (can do as part of Cases 2 and 3)
5. Verify that if rebuildSingleIndexRow throws an exception (can create a scenario using a mock), we "do the right thing". (Is the right thing throw an exception back to the client? If so, I assume we want a retriable exception?)

Reply:
@gjacoby126, I have been testing the read path by completely skipping post index updates manually and letting the verifyRow rebuild the index rows during reads. I will turn this manual testing into an IT test. I will also add a separate test to cover the exception scenario you mentioned.